src/toxencryptsave/encryptsave.rs

```diff
@@ -0,0 +1,165 @@
+    pub key: PrecomputedKey
+}
+
+impl PassKey {
+    pub fn new(passphrase: &[u8]) -> Result<PassKey, KeyDerivationError> {
+        PassKey::with_salt(passphrase, gen_salt())
+    }
+
+    pub fn with_salt(passphrase: &[u8], salt: Salt) -> Result<PassKey, KeyDerivationError> {
+        if passphrase.is_empty() { return Err(KeyDerivationError::Null) };
+
+        let OpsLimit(ops) = OPSLIMIT_INTERACTIVE;
+        let sha256::Digest(passhash) = sha256::hash(passphrase);
+        let maybe_key = PrecomputedKey::from_slice(try!(
+            derive_key(
+                &mut [0; KEY_LENGTH],
+                &passhash,
+                &salt,
+                OpsLimit(ops * 2),
+                MEMLIMIT_INTERACTIVE
+            ).or(Err(KeyDerivationError::Failed))
+        ));
+
+        Ok(PassKey {
+            salt: salt,
+            key: try!(maybe_key.ok_or(KeyDerivationError::Failed))
+        })
+    }
+
+    pub fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>, Encryption> {
+        if data.is_empty() { return Err(Encryption::Null) };
+
+        let mut output = Vec::new();
+        output.extend_from_slice(MAGIC_NUMBER);
+        output.extend_from_slice(&self.salt.0);
+        let nonce = gen_nonce();
+        output.extend_from_slice(&nonce.0);
+        output.append(&mut crypto_core::encrypt_data_symmetric(
+            &self.key,
+            &nonce,
+            data
+        ));
+        Ok(output)
+    }
+
+    pub fn decrypt(&self, data: &[u8]) -> Result<Vec<u8>, Decryption> {
+        if data.is_empty() { return Err(Decryption::Null) };
+        if data.len() <= EXTRA_LENGTH { return Err(Decryption::InvalidLength) };
+        if !is_encrypted(data) { return Err(Decryption::BadFormat) };
+
+        let nonce = try!(Nonce::from_slice(&data[
+            MAGIC_LENGTH+SALT_LENGTH..MAGIC_LENGTH+SALT_LENGTH+NONCEBYTES
+        ]).ok_or(Decryption::BadFormat));
+
+        let output = try!(crypto_core::decrypt_data_symmetric(
+            &self.key,
+            &nonce,
+            &data[MAGIC_LENGTH+SALT_LENGTH+NONCEBYTES..]
+        ).or(Err(Decryption::Failed)));
+
+        Ok(output)
+    }
+}
+
+#[inline]
+pub fn is_encrypted(data: &[u8]) -> bool {
+    data.starts_with(MAGIC_NUMBER)
+}
+
+pub fn pass_encrypt(data: &[u8], passphrase: &[u8]) -> Result<Vec<u8>, Encryption> {
+    try!(PassKey::new(passphrase)).encrypt(data)
+}
+
+pub fn pass_decrypt(data: &[u8], passphrase: &[u8]) -> Result<Vec<u8>, Decryption> {
+    let salt = try!(get_salt(data).ok_or(KeyDerivationError::Failed));
+    try!(PassKey::with_salt(passphrase, salt)).decrypt(data)
+}
+
+pub fn get_salt(data: &[u8]) -> Option<Salt> {
+    if is_encrypted(data)
+        && data.len() >= MAGIC_LENGTH + SALT_LENGTH
+    {
+        let mut salt = [0; SALT_LENGTH];
+        salt.clone_from_slice(&data[MAGIC_LENGTH..MAGIC_LENGTH+SALT_LENGTH]);
+        Some(Salt(salt))
+    } else {
+        None
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum KeyDerivationError {
```

Comment:
* needs docs about what given error means
* should also derive `Copy, Eq, PartialEq`

Same goes for other error enums.